packages/date-fns-jalali/src/date-fns-utils.ts

```diff
@@ -0,0 +1,323 @@
+import addDays from "date-fns-jalali/addDays";
+import addMonths from "date-fns-jalali/addMonths";
+import addYears from "date-fns-jalali/addYears";
+import differenceInMilliseconds from "date-fns-jalali/differenceInMilliseconds";
+import eachDayOfInterval from "date-fns-jalali/eachDayOfInterval";
+import endOfDay from "date-fns-jalali/endOfDay";
+import endOfWeek from "date-fns-jalali/endOfWeek";
+import endOfYear from "date-fns-jalali/endOfYear";
+import format from "date-fns-jalali/format";
+import getHours from "date-fns-jalali/getHours";
+import getSeconds from "date-fns-jalali/getSeconds";
+import getYear from "date-fns-jalali/getYear";
+import getMonth from "date-fns-jalali/getMonth";
+import getMinutes from "date-fns-jalali/getMinutes";
+import isAfter from "date-fns-jalali/isAfter";
+import isBefore from "date-fns-jalali/isBefore";
+import isEqual from "date-fns-jalali/isEqual";
+import isSameDay from "date-fns-jalali/isSameDay";
+import isSameYear from "date-fns-jalali/isSameYear";
+import isSameMonth from "date-fns-jalali/isSameMonth";
+import isSameHour from "date-fns-jalali/isSameHour";
+import isValid from "date-fns-jalali/isValid";
+import dateFnsParse from "date-fns-jalali/parse";
+import setHours from "date-fns-jalali/setHours";
+import setMinutes from "date-fns-jalali/setMinutes";
+import setMonth from "date-fns-jalali/setMonth";
+import setSeconds from "date-fns-jalali/setSeconds";
+import setYear from "date-fns-jalali/setYear";
+import startOfDay from "date-fns-jalali/startOfDay";
+import startOfMonth from "date-fns-jalali/startOfMonth";
+import endOfMonth from "date-fns-jalali/endOfMonth";
+import startOfWeek from "date-fns-jalali/startOfWeek";
+import startOfYear from "date-fns-jalali/startOfYear";
+
+// Locale is not exported from date-fns, so we need to workaround that https://github.com/date-fns/date-fns/issues/932
+import SampleLocale from "date-fns-jalali/locale/fa-jalali-IR";
+import { IUtils } from "@date-io/core/IUtils";
+
+type Locale = typeof SampleLocale;
+
+export default class DateFnsJalaliUtils implements IUtils<Date> {
```

Comment:
Extend class from the default date-fns utils to reduce duplications.

```suggestion
export default class DateFnsJalaliUtils implements IUtils<Date> extends DefaultDateFnsUtils {
```

Reply:
No. Jalali calendar is not only a different format. also, the logic of most of the methods is changed.

For example, the `startOfMonth` function has different logic:

```ts
import { startOfMonth } from 'date-fns-jalali';

startOfMonth(new Date('2020-04-25')) // 2020-04-20 (jalali: 1399-02-01)
```

```ts
import { startOfMonth } from 'date-fns';

startOfMonth(new Date('2020-04-25')) // 2020-04-01 (jalali: 1399-01-13)
```

After I checked, only display methods and 5 other functions had the same logic. rest of them (more than 50 functions) had different logic.

Reply:
Thus we must override all the methods that are different.